File: PHPMasterServerConnect.js
public var masterServerURL : String;
public var gameType : String;
@HideInInspector
public var gameName : String;
@HideInInspector
public var comment : String = "";
public var delayBetweenUpdates : float = 10.0;
private var hostData : HostData[] = null;
public var maxRetries : int = 3;
private var retries :int = 0;


function Awake () {
	var gos = FindObjectsOfType (PHPMasterServerConnect);
  	if (gos.Length > 1) {
		Destroy (gameObject);
	} else {
		DontDestroyOnLoad (this);
	}
}

public function PollHostList () : HostData[] {
	return hostData;
}

public function QueryPHPMasterServer (type : String) {
    url = masterServerURL+"QueryMS.php?gameType="+WWW.EscapeURL(type);
    Debug.Log ("looking for URL " + url);
    www = new WWW (url);
    yield www;

    retries = 0;
    while (www.error != null && retries < maxRetries) {
        retries ++;
        www = new WWW (url);
        yield www;
    }
    if (www.error != null) {
        SendMessage ("OnQueryMasterServerFailed");
        return;
    }

    if (www.text == "") {
    	hostData = null;
        return;
    }
    hosts = www.text.Split (";"[0]);
    hostData = new HostData[hosts.length];
    var index = 0;
    for (var host in hosts) {
        data = host.Split (","[0]);
        hostData[index] = new HostData ();
        hostData[index].ip = new String[1];
        hostData[index].ip[0] = data[0];
        hostData[index].port = parseInt(data[1]);
        hostData[index].useNat = (data[2] == "1");
        hostData[index].guid = data[3];
        hostData[index].gameType = data[4];
        hostData[index].gameName = data[5];
        hostData[index].connectedPlayers = parseInt(data[6]);
        hostData[index].playerLimit = parseInt(data[7]);
        hostData[index].passwordProtected = (data[8] == "1");
        hostData[index].comment = data[9];
        index ++;
    }
}

function RegistrationLoop() {
  while (Network.isServer) {
    yield WaitForSeconds(delayBetweenUpdates);

    var url = masterServerURL+"RegisterHost.php";
    url += "?gameType="+WWW.EscapeURL (gameType);
    url += "&gameName="+WWW.EscapeURL (gameName);
    url += "&comment="+WWW.EscapeURL (comment);
    url += "&useNat="+!Network.HavePublicAddress();
    url += "&connectedPlayers="+(Network.connections.length + 1);
    url += "&playerLimit="+Network.maxConnections;
    url += "&internalIp="+Network.player.ipAddress;
    url += "&internalPort="+Network.player.port;
    url += "&externalIp="+Network.player.externalIP;
    url += "&externalPort="+Network.player.externalPort;
    url += "&guid="+Network.player.guid;
    url += "&passwordProtected="+(Network.incomingPassword != "" ? 1 : 0);
    Debug.Log (url);

    var www = new WWW (url);
    yield www;

    retries = 0;
    while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
        retries ++;
        www = new WWW (url);
        yield www;
    }
    if ((www.error != null || www.text != "succeeded")) {
        SendMessage ("OnUpdateHostFailed");
        return;
    }
  }
}

function OnServerInitialized () {
  while (Network.player.externalPort == 65535) {
    yield WaitForSeconds(1);
  }
  var url = masterServerURL+"RegisterHost.php";
  url += "?gameType="+WWW.EscapeURL (gameType);
  url += "&gameName="+WWW.EscapeURL (gameName);
  url += "&comment="+WWW.EscapeURL (comment);
  url += "&useNat="+!Network.HavePublicAddress();
  url += "&connectedPlayers="+(Network.connections.length + 1);
  url += "&playerLimit="+Network.maxConnections;
  url += "&internalIp="+Network.player.ipAddress;
  url += "&internalPort="+Network.player.port;
  url += "&externalIp="+Network.player.externalIP;
  url += "&externalPort="+Network.player.externalPort;
  url += "&guid="+Network.player.guid;
  url += "&passwordProtected="+(Network.incomingPassword != "" ? 1 : 0);
  Debug.Log (url);
  var www = new WWW (url);
  yield www;

  retries = 0;
  while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
      retries ++;
      www = new WWW (url);
      yield www;
  }
  if ((www.error != null || www.text != "succeeded")) {
      SendMessage ("OnRegisterHostFailed");
      return;
  }

  StartCoroutine (RegistrationLoop());
}

function OnPlayerConnected (player : NetworkPlayer) {
    var url = masterServerURL+"UpdatePlayers.php";
    url += "?gameType="+WWW.EscapeURL (gameType);
    url += "&gameName="+WWW.EscapeURL (gameName);
    url += "&connectedPlayers="+(Network.connections.length + 1);
    Debug.Log ("url " + url);
    var www = new WWW (url);
    yield www;

    retries = 0;
    while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
        retries ++;
        www = new WWW (url);
        yield www;
    }
    if ((www.error != null || www.text != "succeeded")) {
        SendMessage ("OnUpdatePlayersFailed");
        return;
    }
}

function OnPlayerDisconnected (player : NetworkPlayer) {
    var url = masterServerURL+"UpdatePlayers.php";
    url += "?gameType="+WWW.EscapeURL (gameType);
    url += "&gameName="+WWW.EscapeURL (gameName);
    url += "&connectedPlayers="+Network.connections.length;
    Debug.Log ("url " + url);
    var www = new WWW (url);
    yield www;

    retries = 0;
    while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
        retries ++;
        www = new WWW (url);
        yield www;
    }
    if ((www.error != null || www.text != "succeeded")) {
        SendMessage ("OnUpdatePlayersFailed");
        return;
    }
}

function OnDisconnectedFromServer(info : NetworkDisconnection) {
    if (Network.isServer) {
        var url = masterServerURL+"UnregisterHost.php";
        url += "?gameType="+WWW.EscapeURL (gameType);
        url += "&gameName="+WWW.EscapeURL (gameName);
        var www = new WWW (url);
        yield www;

        retries = 0;
        while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
            retries ++;
            www = new WWW (url);
            yield www;
        }
        if ((www.error != null || www.text != "succeeded")) {
            SendMessage ("OnUnregisterHostFailed");
            return;
        }
    }
}

function SetComment (text : String) {
    comment = text; 
}

File: PHPMasterServerConnect.cs
using UnityEngine;
using UnityEngine.Networking;
using System.Collections;

public class PHPMasterServerConnect : MonoBehaviour 
{
	public string masterServerURL = "";
	public string gameType = "";
    [HideInInspector]
	public string gameName = "";
    [HideInInspector]
	public string comment = "";

	public float delayBetweenUpdates = 10.0f;
	private HostData[] hostData = null;

	public int maxRetries = 3;
	private int retries = 0;

	private bool registered = false;
	
	static private PHPMasterServerConnect _instance = null;
	static public PHPMasterServerConnect instance {
		get {
			if (_instance == null) {
				_instance = (PHPMasterServerConnect) FindObjectOfType (typeof(PHPMasterServerConnect));
			}
			return _instance;
		}
	}
	
	void Awake () {
        if (_instance != null) {
            Debug.LogError ("Instance should be null");
            DestroyImmediate (gameObject);
        } else {
		    DontDestroyOnLoad (gameObject);
		}
        _instance = this;
	}

    void OnDestroy () {
        if (registered) {
			// Unregister without the CR
	        string url = masterServerURL+"UnregisterHost.php";
	        url += "?gameType="+WWW.EscapeURL (gameType);
	        url += "&gameName="+WWW.EscapeURL (gameName);
	        new WWW (url);
        }
    }
	
	public HostData[] PollHostList()
	{
		return hostData;
	}

	private bool atOnce = false;
	public void QueryPHPMasterServer ()
	{
		if (!atOnce) 
			StartCoroutine (QueryPHPMasterServerCR ());
	}
	
	private IEnumerator QueryPHPMasterServerCR ()
	{
		atOnce = true;
		string url = masterServerURL+"QueryMS.php?gameType="+WWW.EscapeURL(gameType);
    	// Debug.Log ("looking for URL " + url);
    	WWW www = new WWW (url);
    	yield return www;

    	retries = 0;
	    while (www.error != null && retries < maxRetries) {
	        retries ++;
	        www = new WWW (url);
	        yield return www;
	    }
	    if (www.error != null) {
	        SendMessage ("OnQueryMasterServerFailed");
	    }
	
		if (www.text == "") {
			yield break;
		} else if (www.text == "empty") {
			Debug.LogError ("Got an empty result I didn't expect here!");
	    	hostData = null;
	    } else {
	    	string[] hosts = new string[www.text.Split (";"[0]).Length];
			hosts = www.text.Split (";"[0]);
	    	hostData = new HostData[hosts.Length];
	    	var index = 0;
	    	foreach (string host in hosts) {
				if (host == "") continue;
	        	string[] data = host.Split (","[0]);
	        	hostData[index] = new HostData ();
	        	hostData[index].ip = new string[1];
	        	hostData[index].ip[0] = data[0];
	        	hostData[index].port = int.Parse(data[1]);
	        	hostData[index].useNat = (data[2] == "1");
	        	hostData[index].guid = data[3];
	        	hostData[index].gameType = data[4];
	        	hostData[index].gameName = data[5];
	        	hostData[index].connectedPlayers = int.Parse(data[6]);
	        	hostData[index].playerLimit = int.Parse(data[7]);
	        	hostData[index].passwordProtected = (data[8] == "1");
	        	hostData[index].comment = data[9];
	        	index ++;
	    	}
		}
		atOnce = false;
	}

	public void RegisterHost (string pGameName, string pComment) {
		gameName = pGameName;
		comment = pComment;
		registered = true;
		StartCoroutine (RegistrationLoop ());
	}

	private IEnumerator RegistrationLoop()
	{
		while (registered && NetworkServer.active) {
            yield return StartCoroutine (RegisterHostCR());
    		yield return new WaitForSeconds(delayBetweenUpdates);
		}

		registered = false;
	}

    private IEnumerator RegisterHostCR () {
	    string url = masterServerURL+"RegisterHost.php";
	    url += "?gameType="+WWW.EscapeURL (gameType);
	    url += "&gameName="+WWW.EscapeURL (gameName);
	    url += "&comment="+WWW.EscapeURL (comment);
		url += "&playerLimit="+NetworkManager.singleton.matchSize;
		url += "&connectedPlayers="+NetworkManager.singleton.numPlayers;
		url += "&internalIp="+NetworkManager.singleton.networkAddress;
		url += "&internalPort="+NetworkManager.singleton.networkPort;
		url += "&externalPort="+NetworkManager.singleton.networkPort;
		if (NetworkManager.singleton.serverBindToIP) {
	    	url += "&externalIp="+NetworkManager.singleton.serverBindAddress;
		} else {
	    	url += "&externalIp="+NetworkManager.singleton.networkAddress;
		}
	    Debug.Log (url);
	
	    WWW www = new WWW (url);
	    yield return www;
	
	    retries = 0;
	    while ((www.error != null || www.text != "") && retries < maxRetries) {
	        retries ++;
	        www = new WWW (url);
	        yield return www;
	    }
	    if ((www.error != null || www.text != "")) {
	        SendMessage ("OnRegisterHostFailed");
		}
    }

	public void UnregisterHost ()
	{
		StartCoroutine (UnregisterHostCR ());
	}
	
	private IEnumerator UnregisterHostCR ()
	{
		if (registered) {
	        string url = masterServerURL+"UnregisterHost.php";
	        url += "?gameType="+WWW.EscapeURL (gameType);
	        url += "&gameName="+WWW.EscapeURL (gameName);
			Debug.Log (url);
	        WWW www = new WWW (url);
	        yield return www;
	
	        retries = 0;
	        while ((www.error != null || www.text != "") && retries < maxRetries) {
	            retries ++;
	            www = new WWW (url);
	            yield return www;
	        }
	        if ((www.error != null || www.text != "")) {
	            SendMessage ("OnUnregisterHostFailed");
	        }

			registered = false;
    	}
	}
	
	// TODO update for the new networking
	IEnumerator OnPlayerConnected(NetworkPlayer player)
	{
		string url = masterServerURL+"UpdatePlayers.php";
	    url += "?gameType="+WWW.EscapeURL (gameType);
	    url += "&gameName="+WWW.EscapeURL (gameName);
	    url += "&connectedPlayers="+(Network.connections.Length + 1);
	    Debug.Log ("url " + url);
	    WWW www = new WWW (url);
	    yield return www;
	
	    retries = 0;
	    while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
	        retries ++;
	        www = new WWW (url);
	        yield return www;
	    }
	    if ((www.error != null || www.text != "succeeded")) {
	        SendMessage ("OnUpdatePlayersFailed");
	    }
	}
	
	// TODO update for the new networking
	IEnumerator OnPlayerDisconnected(NetworkPlayer player)
	{
		string url = masterServerURL+"UpdatePlayers.php";
	    url += "?gameType="+WWW.EscapeURL (gameType);
	    url += "&gameName="+WWW.EscapeURL (gameName);
	    url += "&connectedPlayers="+Network.connections.Length;
	    Debug.Log ("url " + url);
	    WWW www = new WWW (url);
	    yield return www;
	
	    retries = 0;
	    while ((www.error != null || www.text != "succeeded") && retries < maxRetries) {
	        retries ++;
	        www = new WWW (url);
	        yield return www;
	    }
	    if ((www.error != null || www.text != "succeeded")) {
	        SendMessage ("OnUpdatePlayersFailed");
	    }
	}
}
